odoo/orm/domains.py

```diff
@@ -0,0 +1,1487 @@
+
+        ```
+        Domain([('a', '=', 5), ('b', '=', 8)])
+        Domain('a', '=', 5) & Domain('b', '=', 8)
+        Domain.AND([Domain('a', '=', 5), *other_domains, Domain.TRUE])
+        ```
+
+        If we have one argument, it is a `Domain`, or a list representation, or a bool.
+        In case we have multiple ones, there must be 3 of them:
+        a field (str), the operator (str) and a value for the condition.
+        """
+        if len(args) > 1:
+            if isinstance(args[0], str):
+                return DomainCondition(*args).checked()
+            # special cases like True/False constants
+            if args == _TRUE_LEAF:
+                return _TRUE_DOMAIN
+            if args == _FALSE_LEAF:
+                return _FALSE_DOMAIN
+            raise TypeError(f"Domain() invalid arguments: {args!r}")
+
+        arg = args[0]
+        if isinstance(arg, Domain):
+            return arg
+        if arg is True or arg == []:
+            return _TRUE_DOMAIN
+        if arg is False:
+            return _FALSE_DOMAIN
+
+        # parse as a list
+        # perf: do this inside __new__ to avoid calling function that return
+        # a Domain which would call implicitly __init__
+        if not isinstance(arg, (list, tuple)):
+            raise TypeError(f"Domain() invalid argument type for domain: {arg!r}")
+        stack: list[Domain] = []
+        try:
+            for item in reversed(arg):
+                if isinstance(item, (tuple, list)) and len(item) == 3:
+                    stack.append(Domain(*item))
+                elif item == DomainAnd.OPERATOR:
+                    stack.append(stack.pop() & stack.pop())
+                elif item == DomainOr.OPERATOR:
+                    stack.append(stack.pop() | stack.pop())
+                elif item == DomainNot.OPERATOR:
+                    stack.append(~stack.pop())
+                elif isinstance(item, Domain):
+                    stack.append(item)
+                else:
+                    raise ValueError(f"Domain() invalid item in domain: {item!r}")
+            # keep the order and simplify already
+            if len(stack) == 1:
+                return stack[0]
+            return Domain.AND(reversed(stack))
+        except IndexError:
+            raise ValueError(f"Domain() malformed domain {arg!r}")
+
+    @classproperty
+    def TRUE(cls) -> Domain:
+        return _TRUE_DOMAIN
+
+    @classproperty
+    def FALSE(cls) -> Domain:
+        return _FALSE_DOMAIN
+
+    @staticmethod
+    def AND(items: Iterable) -> Domain:
+        """Build the conjuction of domains: (item1 AND item2 AND ...)"""
+        return DomainAnd.apply(items)
+
+    @staticmethod
+    def OR(items: Iterable) -> Domain:
+        """Build the disjuction of domains: (item1 OR item2 OR ...)"""
+        return DomainOr.apply(items)
+
+    def __setattr__(self, name, value):
+        if hasattr(self, name):
+            raise RuntimeError("Domain objects are immutable")
```

Comment:
I wonder about the exception type.  I think that `TypeError` would be more adequate in this case, as that's what is recommended for non-supported operations.